### PR TITLE
Added Subarray Sum Java Solution

### PR DESCRIPTION
**Pull request by timothygao8710**

### CODE REVIEW

**Review by dongliu0426 — CHANGES_REQUESTED** [review]
1) use tabs instead of spaces (for indentation)
2) update authors
3) wrap your IO class with Code Snippets

**Comment by HarryWangATX** [comment]
also collapse template

**Comment by dongliu0426** [comment]
Is https://github.com/cpinitiative/usaco-guide/pull/1742/commits/106206de5ca76744b3994d8503c9329361a8ba85 meant for #1753 ?

**Comment by timothygao8710** [comment]
What do you mean by code snippets?

**Comment by dongliu0426** [comment]
@timothygao8710 https://usaco.guide/general/working-mdx/#code-blocks-and-code-snippets

**Comment by timothygao8710** [comment]
Thanks. How do I put my fastio template inside a code snippet though? It's not one of the io methods supported by CodeSnip(and Kattio is too slow).

**Comment by dongliu0426** [comment]
```
//BeginCodeSnip{FastIO}
your fastio class
//EndCodeSnip
```

**Comment by timothygao8710** [comment]
👍

**Comment by dongliu0426** [comment]
> 1. use tabs instead of spaces (for indentation)
> 2. update authors
> 3. wrap your IO class with Code Snippets

huh somehow 1 and 2 were reverted

**Comment by timothygao8710** [comment]
k try again?

**Comment by nathangong** [comment]
Shouldn't there be an implementation for "segtree" in the code?

**Comment by dongliu0426** [comment]
> Shouldn't there be an implementation for "segtree" in the code?

+1. is this tested?

**Comment by timothygao8710** [comment]
Oops. Seems like I accidentally deleted that part when I added the code snippet. Also yes. This is tested(otherwise it wouldn't be a pr?).

File: content/5_Plat/Seg_Ext.mdx [hunk 1]
```diff
@@ -292,6 +292,197 @@ int main() {
...
+					rd.nextInt() - 1,
+					rd.nextInt(),
+					0, nums.length - 1
+			)));
+		}
+ 
+		pw.close();
+	}
+ 
+	static class segtree {
```
Comment (nathangong) [inline content/5_Plat/Seg_Ext.mdx:323]: Use tabs for the segtree class indentations

File: content/5_Plat/Seg_Ext.mdx [hunk 2]
```diff
@@ -292,6 +292,197 @@ int main() {
...
+
+        public segtree left;
```
Comment (nathangong) [inline content/5_Plat/Seg_Ext.mdx:325]: Use camel case for the segtree class

File: solutions/usaco-696.mdx [hunk 3]
```diff
@@ -2,10 +2,102 @@
...
 
-## Merging Indexed Sets
+## Solution 1: Euler Tour
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+Recall from the euler tour module that each subtree is represented as a contiguous interval in it's euler tour. With this knowledge, we can compress the tree into an array. Now, the problem becomes for each node $i$, look for how many elements are less than $tour[i]$ from $i$ to $tout[i]$, where $tout[i]$ is the end of the subtree interval for node $i$. By processing the nodes in decreasing order, this is a problem we can solve in $\mathcal{O}(N\log N)$ time with a binary-indexed tree.
+
+```java
+import java.io.*;
+import java.util.ArrayList;
```
Comment (nathangong) [inline solutions/usaco-696.mdx:15]: just do ```import java.util.*``` for simplicity

File: content/5_Plat/Seg_Ext.mdx (hunk 2 above)
Comment (dongliu0426) [reply content/5_Plat/Seg_Ext.mdx:325]: it's already camel case?

Comment (nathangong) [reply content/5_Plat/Seg_Ext.mdx:325]: Classes should be upper camel case

File: solutions/usaco-696.mdx (hunk 3 above)
Comment (timothygao8710) [reply solutions/usaco-696.mdx:15]: this is actually from a different pr i think

**Review by dongliu0426 — DISMISSED** [review]
https://github.com/cpinitiative/usaco-guide/pull/1742#discussion_r654830012

https://github.com/cpinitiative/usaco-guide/pull/1742#discussion_r654841435
I think he meant UpperCamelCase (`SegTree` instead of `segTree`) ?

File: content/5_Plat/Seg_Ext.mdx [hunk 4]
```diff
@@ -287,11 +287,202 @@ int main() {
...
+					0, nums.length - 1
+			)));
+		}
+ 
+		pw.close();
+	}
+ 
+	static class SegTree {
+
+        public SegTree left;
```
Comment (dongliu0426) [inline content/5_Plat/Seg_Ext.mdx:325]: use tabs instead of spaces

File: solutions/usaco-696.mdx [hunk 5]
```diff
@@ -2,10 +2,102 @@
...
 title: Promotion Counting
-author: Benjamin Qi
+author: Benjamin Qi, Timothy Gao
 ---
 
-## Merging Indexed Sets
+## Solution 1: Euler Tour
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+Recall from the euler tour module that each subtree is represented as a contiguous interval in it's euler tour. With this knowledge, we can compress the tree into an array. Now, the problem becomes for each node $i$, look for how many elements are less than $tour[i]$ from $i$ to $tout[i]$, where $tout[i]$ is the end of the subtree interval for node $i$. By processing the nodes in decreasing order, this is a problem we can solve in $\mathcal{O}(N\log N)$ time with a binary-indexed tree.
```
Comment (dongliu0426) [inline solutions/usaco-696.mdx:11]: textttify `tour` and `tout`

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

**Comment by nathangong** [comment]
I'm not sure if you want the c++ code to also have upper camel case SegTree since the code is using an array to store it instead

File: solutions/usaco-696.mdx (hunk 5 above)
Comment (timothygao8710) [reply solutions/usaco-696.mdx:11]: I have no idea how to do that

Comment (dongliu0426) [reply solutions/usaco-696.mdx:11]: $\texttt{tour}$ $\texttt{tout}$

**Review by dongliu0426 — CHANGES_REQUESTED** [review]
https://github.com/cpinitiative/usaco-guide/pull/1742#discussion_r671619428

https://github.com/cpinitiative/usaco-guide/pull/1742#discussion_r677112985

File: solutions/usaco-696.mdx (hunk 5 above)
Comment (dolphingarlic) [inline solutions/usaco-696.mdx:11]: What are `tour` and `tout`? Also, "Euler" should have a capital e

Comment (dolphingarlic) [inline solutions/usaco-696.mdx:11]: > By processing the nodes in decreasing order

Increasing order?

Comment (dolphingarlic) [inline solutions/usaco-696.mdx:11]: You also don't have to say stuff like "With this knowledge"

**Review by dolphingarlic — CHANGES_REQUESTED** [review]

**Comment by dongliu0426** [comment]
Changes requested have not been made. Free free to create a new PR.

**Pull request closed without merging**